tensorflow/compiler/xla/service/all_reduce_reassociate.cc

```diff
@@ -22,24 +22,88 @@ limitations under the License.
 #include "tensorflow/compiler/xla/service/collective_ops_utils.h"
 #include "tensorflow/compiler/xla/service/hlo_domain_map.h"
 #include "tensorflow/compiler/xla/service/hlo_query.h"
+#include "tensorflow/compiler/xla/service/pattern_matcher.h"
 #include "tensorflow/tsl/platform/errors.h"
 
 namespace xla {
 namespace {
 
+namespace m = match;
+
+bool AreAllreduceKeysEqual(AllReduceKey &key0, AllReduceKey &key1,
```

Comment:
may be for later, but we can potentially make `AllReduceKey` a struct derived from the tuple and add a `IsEqual(bool ignore_element_type)` method to better encapsulate this.

Reply:
yea i was thinking to objectify the allreduceKey since it has a number of members now, it's better to do it as a standalone pr since it affects other passes too.